_pages/dev/ssr/server-side-rendering-contextual-logging.md

```diff
@@ -0,0 +1,244 @@
+
+To implement a custom logger, the developer must create a class that implements the `ExpressServerLogger` interface. Alternatively, they may extend the `DefaultExpressServerLogger` class. 
+
+Once the custom logger class is created, it can be passed to the `logger` property in the configuration object. 
+
+An example of how to do this is shown below:
+
+```ts
+import { ngExpressEngine as engine } from '@nguniversal/express-engine';
+import { NgExpressEngineDecorator } from '@spartacus/setup/ssr';
+import { CustomExpressServerLogger } from './custom-express-server-logger';
+
+[...]
+
+const ngExpressEngine = NgExpressEngineDecorator.get(engine, {
+  // pass custom logger here
+  logger: new CustomExpressServerLogger()
+});
+
+[...]
+```
+
+The following is an example of a custom logger implementation:
+
+```ts
+import { ExpressServerLogger } from '@spartacus/setup/ssr';
+
+export class CustomExpressServerLogger implements ExpressServerLogger {
+
+  //custom implementation of the `log` method
+  log(message: string, context: ExpressServerLoggerContext): void {
+    console.log({ message: `[My Custom LOG]: ${message}`, ...context });
+  }
+
+  [...]
+}
+```
+
+An important aspect of our solution is its extensibility. We have designed our solution to be compatible with enterprise logging libraries used for logging in platform-server such as [Winston](https://www.npmjs.com/package/winston) or [Pino](https://www.npmjs.com/package/pino), making it easy to add their features to our system.
+
+The following is an example of how to implement the custom logger with the `Pino` library:
+
+```ts
+import { DefaultExpressServerLogger, ExpressServerLoggerContext } from '@spartacus/setup/ssr';
+import { pino } from 'pino';
+
+class CustomLogger extends DefaultExpressServerLogger {
+  private logger = pino({ customLevels: { log: 15 }, level: 'log' }); // as there is no `log` level in pino, we need to define it here;
+
+  [...]
+
+  // custom implementation of the `log` method
+  log(message: string, context: ExpressServerLoggerContext): void {
+    // custom log level is used here
+    this.logger.log(this.handleContext(context), message);
+  }
+
+  warn(message: string, context: ExpressServerLoggerContext): void {
+    //built-in log level is used here
+    this.logger.warn(this.handleContext(context), message);
+  }
+
+  [...]
+
+  // method to handle the context before it is passed to the logger
+  protected handleContext(context: ExpressServerLoggerContext) {
+    const { request, ...restOfContext } = context;
+    // `mapRequest` is used to map the request to the log message. This is a method provided in the `DefaultExpressServerLogger` class.
+    // NOTE: for some cases request object is not passed to the context, so we need to check if it exists
+    const mappedRequest = request ? this.mapRequest(request) : undefined;
+    return { ...restOfContext, request: mappedRequest };
+  }
```

Comment:
1. First of all, this code looks very complex, lots of comments about gotchas. TBH as a customer, I'd love to have something like this:
1) If I'm already extending `DefaultExpressServerLogger`, I shouldn't have to write 6 methods(!) (log, info, ..., error) and some context mapper just in order to add some additional context.
2) It should suffice that I override 1 method to reuse the default mapper and populate to the final context some small addiditional portion of data.

1. > // NOTE: for some cases request object is not passed to the context, so we need to check if it exists
It looks bad that customers need to know gotchas for just a simple customization use case. Maybe we made a mistake by passing ssrOptions to this concrete logger without `request`, even though it's a required property of the interface. 

1. Let's present here some very simple customization use case from real life - add to context a value of some HTTP header.

Reply:
Thank you for pointing this out! Yes, during writing an example of how to extend the solution provided by us, I realised it's not that straightforward and needs to be simplified. From Customer's POV provided example may look complicated showing some weak (?) points of our code. 

Fix for such an issue has been addressed in the following PR: https://github.com/SAP/spartacus/pull/17578